src/AspNetCore/Src/applicationinfo.cpp

```diff
@@ -0,0 +1,330 @@
+    _In_ ASPNETCORE_CONFIG   *pConfiguration,
+    _In_ FILE_WATCHER        *pFileWatcher
+)
+{
+    HRESULT hr = S_OK;
+
+    DBG_ASSERT(pConfiguration);
+    DBG_ASSERT(pFileWatcher);
+
+    m_pConfiguration = pConfiguration;
+    hr = m_applicationInfoKey.Initialize(pConfiguration->QueryConfigPath()->QueryStr());
+    if (FAILED(hr))
+    {
+        goto Finished;
+    }
+
+    if (m_pFileWatcherEntry == NULL)
+    {
+        m_pFileWatcherEntry = new FILE_WATCHER_ENTRY(pFileWatcher);
+        if (m_pFileWatcherEntry == NULL)
+        {
+            hr = E_OUTOFMEMORY;
+            goto Finished;
+        }
+    }
+
+    UpdateAppOfflineFileHandle();
+
+Finished:
+    return hr;
+}
+
+HRESULT
+APPLICATION_INFO::StartMonitoringAppOffline()
+{
+    HRESULT hr = S_OK;
+    if (m_pFileWatcherEntry != NULL)
+    {
+        hr = m_pFileWatcherEntry->Create(m_pConfiguration->QueryApplicationPhysicalPath()->QueryStr(), L"app_offline.htm", this, NULL);
+    }
+    return hr;
+}
+
+VOID
+APPLICATION_INFO::UpdateAppOfflineFileHandle()
+{
+    STRU strFilePath;
+    UTILITY::ConvertPathToFullPath(L".\\app_offline.htm", m_pConfiguration->QueryApplicationPhysicalPath()->QueryStr(), &strFilePath);
+    APP_OFFLINE_HTM *pOldAppOfflineHtm = NULL;
+    APP_OFFLINE_HTM *pNewAppOfflineHtm = NULL;
+
+    if (INVALID_FILE_ATTRIBUTES == GetFileAttributes(strFilePath.QueryStr()) && GetLastError() == ERROR_FILE_NOT_FOUND)
+    {
+        m_fAppOfflineFound = FALSE;
+    }
+    else
+    {
+        m_fAppOfflineFound = TRUE;
+        pNewAppOfflineHtm = new APP_OFFLINE_HTM(strFilePath.QueryStr());
+
+        if (pNewAppOfflineHtm != NULL)
+        {
+            if (pNewAppOfflineHtm->Load())
+            {
+                //
+                // loaded the new app_offline.htm
+                //
+                pOldAppOfflineHtm = (APP_OFFLINE_HTM *)InterlockedExchangePointer((VOID**)&m_pAppOfflineHtm, pNewAppOfflineHtm);
+
+                if (pOldAppOfflineHtm != NULL)
+                {
+                    pOldAppOfflineHtm->DereferenceAppOfflineHtm();
+                    pOldAppOfflineHtm = NULL;
+                }
+            }
+            else
+            {
+                // ignored the new app_offline file because the file does not exist.
+                pNewAppOfflineHtm->DereferenceAppOfflineHtm();
+                pNewAppOfflineHtm = NULL;
+            }
+        }
+
+        // recycle the application
+        if (m_pApplication != NULL)
+        {
+            m_pApplication->ShutDown();
+            m_pApplication->DereferenceApplication();
+            m_pApplication = NULL;
+        }
+    }
+}
+
+HRESULT
+APPLICATION_INFO::EnsureApplicationCreated()
+{
+    HRESULT hr = S_OK;
+    BOOL    fLocked = FALSE;
+    APPLICATION* pApplication = NULL;
+    STACK_STRU(struFileName, 300);  // >MAX_PATH
+    STRU        hostFxrDllLocation;
+
+    if (m_pApplication != NULL)
+    {
+        goto Finished;
+    }
+
+    FindRequestHandlerAssembly();
```

Comment:
The requesthandler.dll will have two functions exposed on it; CreateApplication (on a per application basis) and CreateRequestHandler (per request basis). Calling FindRequestHandlerAssembly will call into hostfxr to find it.